Add mapping for Doctrine core entities on front environment

jolelievre opened the pull request:
<!-----------------------------------------------------------------------------
Thank you for contributing to the PrestaShop project! 

Please take the time to edit the "Answers" rows below with the necessary information.

Check out our contribution guidelines to find out how to complete it:
https://devdocs.prestashop.com/1.7/contribute/contribution-guidelines/#pull-requests
------------------------------------------------------------------------------>

| Questions     | Answers
| ------------- | -------------------------------------------------------
| Branch?       | 1.7.7.x
| Description?  | The Doctrine core entities were not usable in the front container (built by PrestaShop core). This PR fixes this issue by bringing a few chances:<br>- extract doctrine config into dedicated yaml<br>- introduce new YAML config files dedicated to the front<br>- introduce `ConfigYamlLoader` class to load this config in `DoctrineExtension`<br>- integrates a fix originally coming from this PR https://github.com/PrestaShop/PrestaShop/pull/17706 which fixes `Module::get` and new methods like `Module::hasSymfonyContainer` and `Module::hasLegacyContainer` to better detect the environment/container
| Type?         | bug fix
| Category?     | CO
| BC breaks?    | no
| Deprecations? | no
| Fixed ticket? | Fixes #19015
| How to test?  | See the issue to reproduce the bug with the example module, with this PR the entities now work correctly and the quotes are displayed on the home page (you should remove your cache after switching to the PR as the container needs to be rebuilt)

<!-- Click the form's "Preview" button to make sure the table is functional in GitHub. Thank you! -->

<!-- Reviewable:start -->
---
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/prestashop/prestashop/19016)
<!-- Reviewable:end -->

Note: To test this PR you will need this example module https://github.com/PrestaShop/example-modules/pull/20

# Install instruction

- `clone` the demomodule in your `modules` folder (be careful that the repository contains multiple folder, you only need the `demodoctrine` folder)
- `checkout` the appropriate branch for the module `demodoctrine`
- run `composer install` in the module folder
- in the *PrestaShop* root folder make sure you are on the correct branch `doctrine-core-entities`
- run the `rm -fR var/cache/*` command to be sure all cache is empty
- you should now be able to install the module without issue (in the BO or using the CLI command `php bin/console prestashop:module install demodoctrine`

jolelievre committed:
Extract doctrine config in a dedicated yaml shared in each config that needs it, introduce new front YAML config files\
, add ConfigYamlLoader to load them Add core mapping for front doctrine environment

Progi1984 commented on src/PrestaShopBundle/DependencyInjection/Compiler/CoreDoctrineCompilerPass.php:
```diff
@@ -0,0 +1,78 @@
+<?php
+/**
+ * 2007-2019 PrestaShop and Contributors
```
Two things : the copyright date & `declare(strict_types=1);`

PierreRambaud replied:
It's for the 1.7.6.x we can't use the strict types

PierreRambaud commented on src/PrestaShopBundle/DependencyInjection/Compiler/CoreDoctrineCompilerPass.php:
```diff
@@ -0,0 +1,78 @@
+     *
+     * @return DoctrineOrmMappingsPass
+     */
+    private function createAnnotationMappingDriver($moduleNamespace, $moduleEntityDirectory, $modulePrefix)
```
Do you tried with https://symfony.com/doc/current/reference/configuration/doctrine.html#mapping-entities-outside-of-a-bundle ?

jolelievre replied:
I didn't know about that! Smart!
I used this and removed the compiler pass, I only added it in the `DoctrineExtension` because we only need this in the front container

PierreRambaud replied:
You can do it in a yaml file :) instead of editing the php code

jolelievre replied:
Same difference, and I need to add it to the core config which is already parsed into an array So there would be the few interest two parse two yaml files and merge them partially 🤔 I wonder where I would place this file

PierreRambaud reviewed (dismissed):
We should use the config.yml the same place where the `doctrine` component is defined.
It's easier for anyone to improve the configuration if it's stored at the same place.
PIng @eternoendless

jolelievre commented:
The thing is this is a special case for our light container, only available in legacy environment The generic doctrine config for symfony environment is in `app/config/config.yml`

We could add this no-bundle config in it but it would make it redundant in symfony env because the mapping is automatically managed by the bundle I tried it, it **seems** to work fine but I'm worried about potential side effects, and it forces us to add useless config in the Symfony config just to satisfy our custom front container..

An alternative solution could be to add a `app/config/config_front.yml` config file that would import the config file By the way this makes me realize we never load the config depending on the env, meaning doctrine doesn't use any cache in front for now

eternoendless commented:
I agree on the settings approach as we discussed, but I think this PR should target 1.7.7.x

jolelievre committed:
Optimize and fix the Module function related to container and services which did not work in CLI environment

jolelievre commented on src/Adapter/Container/DoctrineBuilderExtension.php:
```diff
@@ -40,25 +41,30 @@
+        $fileLocator = new FileLocator($configDirectories);
+
+        $configLoader = new ConfigYamlLoader($fileLocator);
+        $configPath = sprintf('config_front_%s.yml', $this->environment->getName());
```
@PierreRambaud @eternoendless I'm not sure about the `_front` prefix because it is used for legacy containers, which is used on legacy context for both front and admin

So I see two solutions:
- use `_legacy` prefix
- add three new files with `_admin` prefix which will be the same as the front ones, but it complicates the config files even more

Note: this legacy admin container is supposed to be removed in 178 because only the Symfony container will be available in BO, so maybe we can accept this for one version Starting 178 the `_front` files and legacy container will only be used for front

PierreRambaud replied:
:+1: for the `_legacy` prefix and the note :)

jolelievre committed:
Rename config_front files into config_legacy

jolelievre committed:
No exception catch in Module::get method as it actually created side effects

Progi1984 commented on src/Adapter/Container/DoctrineBuilderExtension.php:
```diff
@@ -40,25 +41,30 @@
     public function build(ContainerBuilder $container)
     {
-        $yamlParser = new YamlParser($this->cacheDir);
+        $configDirectories = [_PS_ROOT_DIR_ . '/app/config'];
```
```suggestion
        $configDirectories = [$container->getParameter('kernel.project_dir') . '/app/config'];
```

jolelievre replied:
Yep you're right!

Progi1984 reviewed (commented):
A small feedback

jolelievre committed:
Use kernerl.project_dir instead of legacy const

eternoendless commented on app/config/config_legacy.yml:
```diff
@@ -0,0 +1,12 @@
+imports:
```
Consider adding a header comment in this file explaining it

eternoendless commented on app/config/doctrine.yml:
```diff
@@ -0,0 +1,34 @@
+        user:     "%database_user%"
+        password: "%database_password%"
+        charset:  utf8mb4
+        # if using pdo_sqlite as your database driver:
+        #   1. add the path in parameters.yml
+        #     e.g. database_path: "%kernel.root_dir%/data/data.db3"
+        #   2. Uncomment database_path in parameters.yml.dist
+        #   3. Uncomment next line:
+        #     path:     "%database_path%"
```
I suppose these lines can go away

eternoendless commented on app/config/doctrine.yml:
```diff
@@ -0,0 +1,34 @@
+
+  orm:
+    auto_generate_proxy_classes: "%kernel.debug%"
+    # naming_strategy: doctrine.orm.naming_strategy.underscore
```
I suppose this line can be deleted

eternoendless commented on classes/module/Module.php:
```diff
@@ -3432,33 +3438,54 @@ public function isAdminLegacyContext()
+     *  - Legacy environment admin: container generated by PrestaShop core for admin controllers
+     *  - Legacy environment front: container generated by PrestaShop core for front controllers
+     *
+     * @return ContainerBuilder|ContainerInterface|null
+     */
+    public function getContainer()
+    {
+        if (null === $this->container) {
+            $finder = new ContainerFinder($this->context);
+            try {
+                $this->container = $finder->getContainer();
+            } catch (ContainerNotFoundException $e) {
+                return null;
+            }
```
"A method should do what it's meant to do or throw an exception". Avoid mixed return types, because it's error prone and doesn't add value.

Consider this code:

```php
$this->getContainer()->get('some.service'); // Error: call to member function on null
```

It's best to remove the try and let consumers handle exceptions:
```suggestion
     * @return ContainerInterface
     *
     * @throws ContainerNotFoundException
     */
    public function getContainer(): ContainerInterface
    {
        if (null === $this->container) {
            $finder = new ContainerFinder($this->context);
            $this->container = $finder->getContainer();
```

jolelievre replied:
Ok fine by me, I actually introduced a similar check in the `get` method but had similar trouble so it's logical to remove it here as well

eternoendless commented on classes/module/Module.php:
```diff
@@ -3432,33 +3438,54 @@ public function isAdminLegacyContext()
     public function isSymfonyContext()
     {
-        return !$this->isAdminLegacyContext() && defined('_PS_ADMIN_DIR_');
+        // Check that a container is available but not generated by PrestaShop
+        $container = $this->getContainer();
+
+        return null !== $container && !($container instanceof LegacyContainerInterface);
```
I think we should keep this as it was. It's the uses that must be updated to perform duck-typing (aka "is the container present?")

jolelievre replied:
But having a container doesn't tell if it's a full filled container from Symfony (with CQRS, bundles, and so on)
Or a legacy one with restricted services and features

So we could leave `isSymfonyContext` untouched, but add a new method to easily test which kind of container is present

Although I think the behaviour of this method has already been modified since 175, because some native modules like ps_linklist or productcomments can't be installed via CLI any more, so something changed

jolelievre replied:
@eternoendless I reverted the changes on `isSymfonyContext` so no more changes nor BC break
And instead I added two methods `hasSymfonyContainer|hasLegacyContainer` which will allow to detect the context/container more easily

eternoendless replied:
I don't think we should add more "maybe useful someday" methods in a class which already does too much.

jolelievre replied:
Ok I'll remove this one, and explain this in the PHPDoc of `getContainer`

eternoendless commented on src/Adapter/Container/ConfigYamlLoader.php:
```diff
@@ -0,0 +1,105 @@
+ */
+declare(strict_types=1);
+
+namespace PrestaShop\PrestaShop\Adapter\Container;
```
Why is this in Adapter?

jolelievre replied:
Mainly because it's used by `DoctrineBuilderExtension` only, but this one might be moved in a different namespcae as it's quite generic

eternoendless replied:
Okay

eternoendless commented on src/Adapter/Container/ConfigYamlLoader.php:
```diff
@@ -0,0 +1,105 @@
+/**
+ * ConfigYamlLoader is used to load YAML config files and get the config as array.
+ */
+class ConfigYamlLoader extends FileLoader
```
This class looks pretty generic. Doesn't symfony already handle this stuff?

jolelievre replied:
Not for config files only, they have `Symfony\Component\DependencyInjection\Loader\YamlFileLoader` but it's meant to loader services and needs a container as input
The `parseImports` from this class is inspired by their `YamlFileLoader` but I needed something more simple that simply loads a config YAML files along with its imported files, and I couldn't find any class that did this (based on a search on the abstract `FileLoader` class)

jolelievre replied:
Maybe we could move this in `PrestaShopBundle\DependencyInjection\Config` ?

eternoendless commented on src/Adapter/Container/DoctrineBuilderExtension.php:
```diff
@@ -27,8 +27,9 @@
 namespace PrestaShop\PrestaShop\Adapter\Container;
```
I wonder why this is in Adapter... shouldn't it be in the bundle?

jolelievre replied:
Because it is used by our `ContainerBuilder` which namespace is in Adapter

This class is not an extension as in a Symfony extension, it's a custom extension used by our `ContainerBuilder` only, which implements our `PrestaShop\PrestaShop\Adapter\Container\ContainerBuilderExtensionInterface` So it shouldn't be in the bundle as it's not related to Symfony framework and can't be used by it

An extra pre layer was required additionally to the `CompilerPassInterfaces` because some actions need to be done before the container is built and start processing the compiler passes So I introduced this extension notion to manage Doctrine, but from the recent modification from this PR I actually wonder if we could have used the Symfony extension system instead

Also having a centralized/common config parsed by the container builder and passed to extensions might have been nice (that's the Symfony principle), but it represents a more in depth refacto

eternoendless replied:
Oh I see.

eternoendless commented on src/Adapter/Container/DoctrineBuilderExtension.php:
```diff
@@ -40,25 +41,30 @@
     public function build(ContainerBuilder $container)
     {
-        $yamlParser = new YamlParser($this->cacheDir);
+        $configDirectories = [$container->getParameter('kernel.project_dir') . '/app/config'];
```
```suggestion
        $configDirectories = [$container->getParameter('kernel.root_dir') . '/config'];
```
Although this path may be better set up as a parameter instead of hard coded?

jolelievre replied:
Good idea for `kernel.root_dir` however I'm not sure it's worth adding a parameter for the config directory, it will remain here by convention So letting the extension compute the path based on the `kernel.root_dir` seems enough to me

jolelievre committed:
Improvements from feedback

jolelievre commented:
@eternoendless I addressed some of your feedback, and commented the others And as the `ps_linklist` install via CLI didn't work any more I proposed a fix on the module https://github.com/PrestaShop/ps_linklist/pull/86

(Actually it was not related to any modif in the module class, it only worked in 175 because it didn't use the container or service back then)

jolelievre committed:
Revert BC break and add two checkers for container

jolelievre committed:
Change namespace for ConfigYamlLoader

eternoendless commented on classes/module/Module.php:
```diff
@@ -3436,30 +3442,70 @@ public function isSymfonyContext()
     }
+
+    /**
+     * Return the container depending on the context:
+     *  - Symfony environment: container generated by the kernel
+     *  - Legacy environment admin: container generated by PrestaShop core for admin controllers
+     *  - Legacy environment front: container generated by PrestaShop core for front controllers
```
Phpdoc is meant to explain the _what_, not the _how_. If the _how_ is valuable information, keep it in the delegate service or in the method body.

```suggestion
     * Returns the service container.
```

jolelievre replied:
But here I wanted to explain what can be returned, because depending on which type of container is returned the services available are not the same So maybe it's just about modifying the description of each environment/container

eternoendless commented on classes/module/Module.php:
```diff
@@ -3436,30 +3442,70 @@ public function isSymfonyContext()
+     *
+     * @return ContainerBuilder|ContainerInterface|null
+     */
+    public function getContainer()
```
I don't see  why this should be public

```suggestion
    protected function getContainer()
```

eternoendless replied:
Alternatively

```suggestion
    protected function getContainer(): ContainerInterface
```

jolelievre replied:
If we remove the `hasSymfonyContainer` `hasLegacyContainer` then this one is useful as public to check the container type manually

Besides it's also useful to get a parameter: `$this->getContainer()->getParameter()` unless we introduce some new shortcut `Module::getParameter`

eternoendless replied:
Ok then

eternoendless commented on classes/module/Module.php:
```diff
@@ -3436,30 +3442,70 @@ public function isSymfonyContext()
+     *  - Legacy environment admin: container generated by PrestaShop core for admin controllers
+     *  - Legacy environment front: container generated by PrestaShop core for front controllers
+     *
+     * @return ContainerBuilder|ContainerInterface|null
```
I suppose all containers implement `ContainerInterface`? In that case you can just hint that interface. And add it as a return type.

eternoendless replied:
BTW this return type is also wrong in `ContainerFinder` which will never return null.

jolelievre replied:
I removed `ContainerBuilder` indeed, but I put `LegacyContainerInterface` instead to emphasize the fact that two king of containers can be returned (Symfony and legacy)

eternoendless replied:
They still belong to the same class hierarchy

eternoendless commented on classes/module/Module.php:
```diff
@@ -3436,30 +3442,70 @@ public function isSymfonyContext()
-        if ($this->context->controller instanceof Controller) {
-            return $this->context->controller->get($serviceName);
+        if (null !== $container) {
+            return $container->get($serviceName);
```
I find this implementation a little awkward. Especially since `getContainer` will never return null.

Here's how it can be simplified:

```php
try {
	$container = $this->getContainer();
}
catch (ContainerNotFoundException $e) {
	return false;
}

return $container->get($serviceName);
```

jolelievre replied:
You're right I'll also correct ContainerFinder doc

eternoendless commented on src/Adapter/ContainerBuilder.php:
```diff
@@ -197,7 +203,7 @@ private function loadDoctrineAnnotationMetadata()
-     * @throws \Exception
+     * @throws Exception
      */
     private function loadServicesFromConfig(SfContainerBuilder $container)
```
Shouldn't all references to `SfContainerBuilder` be  changed as well?

jolelievre replied:
Yes, you're right

jolelievre committed:
Simplify Module and clear ContainerBuilder

eternoendless commented on classes/module/Module.php:
```diff
@@ -3436,29 +3441,48 @@ public function isSymfonyContext()
+     * If you need to detect which kind of container you are using you can check if it is an instance of LegacyContainerInterface,
+     * which means it's a legacy/light container.
+     *
+     * @return LegacyContainerInterface|ContainerInterface
```
No need to specify subclasses in types

```suggestion
     * @return ContainerInterface
```

jolelievre replied:
It was to emphasize the two different types for type checking, but since I already explained it in the comment I can remove it

eternoendless commented on classes/module/Module.php:
```diff
@@ -3436,29 +3441,48 @@ public function isSymfonyContext()
+     *
+     * @throws ContainerNotFoundException
+     */
+    public function getContainer()
```
Why not set the return type?

```suggestion
    public function getContainer(): ContainerInterface
```

jolelievre replied:
Done

jolelievre committed:
Improve getContainer return type
